pkg/apis/serving/v1beta1/trained_model.go

```diff
@@ -0,0 +1,48 @@
+package v1beta1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+)
+
+type TrainedModel struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              TrainedModelSpec         `json:"spec,omitempty"`
+	Status            TrainedModelDeployStatus `json:"status,omitempty"`
+}
+type TrainedModelSpec struct {
+	// Required field for parent inference service
+	InferenceService string `json:"inferenceService"`
+	// Predictor model spec
+	PredictorModel ModelSpec `json:"predictorModel"`
+	// Explainer model spec
+	ExplainerModel ModelSpec `json:"explainerModel,omitempty"`
+}
+type ModelSpec struct {
+	// Storage URI for the model repository
+	StorageURI string `json:"storageUri"`
+	// Default to latest
+	ModelVersionPolicy ModelVersionPolicy `json:"modelVersionPolicy,omitempty"`
+	// ML framework name
+	// +kubebuilder:webhooks:Enum={"tensorflow","pytorch","sklearn","onnx","xgboost", "custom"}
```

Comment:
I think we should avoid restricting this enum to maintain extensibility.

Reply:
Isn't that what `custom` is for? or what kind of extensibility are you referring to?

Reply:
Could this be a path toward named  "custom" types? You can imagine implementing the KFServer interface for a proprietary format that has reusability and can be configured similarly to the name->image mapping the other frameworks have.

Reply:
I've heard from multiple users that they want to be able to extend KFServing without being forced into the "custom" bucket. I.e., they want their own first class models for their organization. One of those users was bloomberg :)

Reply:
I can see that someone want a "MyAwesomeInternalFramework" here, I can relax the enum check. Since we are on this topic, how do you think we can manage the mapping between machine learning framework and model server?

Reply:
As long as we're using explicit assignment (i.e. model specifies its model server) we don't need any intelligent mapping. With automatic assignment, I'd expect model servers would advertise supported frameworks, perhaps using a whitelist or regex? Then some process (kfserving controller or kfserving sidecar/agent) would query the model server and do assignment.

Reply:
exactly the model server would advertise the supported framework. e.g triton supports pytorch, tensorflow and onnx, so `TrainedModel` webhook would reject sklearn model if user tries to deploy that onto a parent triton inference server.

Reply:
One note on this, ideally we want to make sure we have clean separation between control and data plane components. Ideally the control plane should communicate *only* through the k8s master apis. Happy to go into this further, but it has huge ramifications for the portability of the control plane (i.e. running on another network fabric or even managed cloud provider).

This implies that the InferenceService would need to contain "supported frameworks" in it's status. Perhaps this value can be set by the KFServing Agent? 

KFServing Agent calling K8s Masters (good)
KFServing Controller calling K8s Masters (good)
KFServing Agent communicating with KFServing Controller (bad)

Reply:
I totally agree that the data pane and control pane should be separate. But why is KFServing Agent communicating with KFServing Controller is bad?

Reply:
It's about network fabric. In locked down cluster scenarios, the kubernetes API may be the only whitelisted network ingress/egress between the control and data plane. Imagine a scenario where the KFServing controller isn't running in the cluster. Perhaps it's running on another cluster, or another environment entire (like AWS ECS).

Reply:
Can `supported frameworks` advertised in the KFServing controller configmap?

Reply:
Yeah, I was going to add that in the configMap. Distracted by too many meetings!

Reply:
@ellistarn @yuzisun @kelaban @ifilonenko What's your opinion about introducing a TrainedModelBinding CRD? Our current approach requires users to create an InferenceService first and then create TrainedModel CR to load models into one of their InferenceServices.

I think introducing the TrainedModelBinding CRD can decouple model and model server.  The advantages are:
1) Users can create TrainedModel CR without creating a model server first.
2) An TrainedModel CR can be loaded into multiple model servers, this approach can reduce duplicate TrainedModels.

Reply:
I think `TrainedModelBinding` blurred the line of model management system and KFServing, KFServing ideally shouldn't be in a position of staging the trained model which isn't ready for deployment yet imo.

Reply:
I agree that KFServing should not be responsible for models that aren't ready for deployment. I can imagine that there will be a lot of models only stored in a storage system but do not get deployed out.  

Users should only create a TrainedModel CR when they want to deploy it. Having TrainedModelBinding allows users to deploy the same model to multiple model servers and create TrainedModel CR without running a model server first.

I do think it should be a phase 2 things, we don't have to add it now but want to get people to think about it.

Reply:
I think bindings are too far. Are you worried about duplicating the model's fields? I'm worried about the blast radius of all the models in production if it's served in 4 places and 1 edits the model and the bindings push the change to all consumers.

Reply:
I see, I got your concern about binding. I was thinking from a user's point of view. Users may have a ML pipeline that process data -> train model -> save model to S3 -> create TrainedModel CR. It seems to me that one model can be deployed to multiple model servers and we shouldn't require users to have a model server running before they can create TrainedModel CRs. 

We can think about it more, it does not need to be part of this PR.